File: lib/features/saliks/domain/entities/salik.dart
import 'approval_status.dart';
import '../../../../core/utils/text_field_merge.dart';

class Salik {
  const Salik({
    required this.salikId,
    required this.name,
    required this.fatherName,
    required this.mobileNumber,
    required this.whatsappNumber,
    required this.areaId,
    this.address = '',
    required this.genderId,
    required this.bazamId,
    required this.khanqahId,
    required this.salikCategoryId,
    required this.dateOfBaith,
    required this.referenceName,
    required this.referenceMobile,
    required this.isNafiAsbat,
    required this.isSahibEMehfil,
    required this.profilePicture,
    required this.createdDate,
    required this.modifiedDate,
    required this.isActive,
    this.nafiZikrId = '',
    this.notes,
    this.addedByUid = '',
    this.addedByName = '',
    this.approvalStatus = ApprovalStatus.approved,
    this.approvedByUid = '',
    this.approvedByName = '',
    this.approvedAt = '',
  });

  final String salikId;
  final String name;
  final String fatherName;
  final String mobileNumber;
  final String whatsappNumber;
  final String areaId;
  final String address;
  final String genderId;
  final String bazamId;
  final String khanqahId;
  final String salikCategoryId;
  final String dateOfBaith;
  final String referenceName;
  final String referenceMobile;
  final bool isNafiAsbat;
  final bool isSahibEMehfil;
  final String nafiZikrId;
  final String profilePicture;
  final String createdDate;
  final String modifiedDate;
  final bool isActive;
  final String? notes;
  final String addedByUid;
  final String addedByName;
  final ApprovalStatus approvalStatus;
  final String approvedByUid;
  final String approvedByName;
  final String approvedAt;

  bool get isApproved => approvalStatus == ApprovalStatus.approved;
  bool get isPending => approvalStatus == ApprovalStatus.pending;
  bool get isRejected => approvalStatus == ApprovalStatus.rejected;

  Salik copyWith({
    String? salikId,
    String? name,
    String? fatherName,
    String? mobileNumber,
    String? whatsappNumber,
    String? areaId,
    String? address,
    String? genderId,
    String? bazamId,
    String? khanqahId,
    String? salikCategoryId,
    String? dateOfBaith,
    String? referenceName,
    String? referenceMobile,
    bool? isNafiAsbat,
    bool? isSahibEMehfil,
    String? nafiZikrId,
    String? profilePicture,
    String? createdDate,
    String? modifiedDate,
    bool? isActive,
    String? notes,
    String? addedByUid,
    String? addedByName,
    ApprovalStatus? approvalStatus,
    String? approvedByUid,
    String? approvedByName,
    String? approvedAt,
  }) {
    return Salik(
      salikId: salikId ?? this.salikId,
      name: name ?? this.name,
      fatherName: fatherName ?? this.fatherName,
      mobileNumber: mobileNumber ?? this.mobileNumber,
      whatsappNumber: whatsappNumber ?? this.whatsappNumber,
      areaId: areaId ?? this.areaId,
      address: address ?? this.address,
      genderId: genderId ?? this.genderId,
      bazamId: bazamId ?? this.bazamId,
      khanqahId: khanqahId ?? this.khanqahId,
      salikCategoryId: salikCategoryId ?? this.salikCategoryId,
      dateOfBaith: dateOfBaith ?? this.dateOfBaith,
      referenceName: referenceName ?? this.referenceName,
      referenceMobile: referenceMobile ?? this.referenceMobile,
      isNafiAsbat: isNafiAsbat ?? this.isNafiAsbat,
      isSahibEMehfil: isSahibEMehfil ?? this.isSahibEMehfil,
      nafiZikrId: nafiZikrId ?? this.nafiZikrId,
      profilePicture: profilePicture ?? this.profilePicture,
      createdDate: createdDate ?? this.createdDate,
      modifiedDate: modifiedDate ?? this.modifiedDate,
      isActive: isActive ?? this.isActive,
      notes: notes ?? this.notes,
      addedByUid: addedByUid ?? this.addedByUid,
      addedByName: addedByName ?? this.addedByName,
      approvalStatus: approvalStatus ?? this.approvalStatus,
      approvedByUid: approvedByUid ?? this.approvedByUid,
      approvedByName: approvedByName ?? this.approvedByName,
      approvedAt: approvedAt ?? this.approvedAt,
    );
  }

  factory Salik.fromMap(Map<String, dynamic> map, {String? id}) {
    return Salik(
      salikId: id ?? map['salikId'] as String? ?? '',
      name: readUnifiedText(
        map,
        key: 'name',
        legacyPrimaryKey: 'nameEnglish',
        legacySecondaryKey: 'nameUrdu',
      ),
      fatherName: readUnifiedText(
        map,
        key: 'fatherName',
        legacyPrimaryKey: 'fatherNameEnglish',
        legacySecondaryKey: 'fatherNameUrdu',
      ),
      mobileNumber: map['mobileNumber'] as String? ?? '',
      whatsappNumber: map['whatsappNumber'] as String? ?? '',
      areaId: (map['areaId'] as String? ?? map['cityId'] as String?)?.trim() ?? '',
      address: map['address'] as String? ?? '',
      genderId: map['genderId'] as String? ?? 'Male',
      bazamId: map['bazamId'] as String? ?? '',
      khanqahId: map['khanqahId'] as String? ?? '',
      salikCategoryId: map['salikCategoryId'] as String? ?? '',
      dateOfBaith: map['dateOfBaith'] as String? ?? '',
      referenceName: map['referenceName'] as String? ?? '',
      referenceMobile: map['referenceMobile'] as String? ?? '',
      isNafiAsbat: map['isNafiAsbat'] as bool? ?? false,
      isSahibEMehfil: map['isSahibEMehfil'] as bool? ?? false,
      nafiZikrId: map['nafiZikrId'] as String? ?? '',
      profilePicture: map['profilePicture'] as String? ?? '',
      createdDate: map['createdDate'] as String? ?? '',
      modifiedDate: map['modifiedDate'] as String? ?? '',
      isActive: map['isActive'] as bool? ?? true,
      notes: map['notes'] as String?,
      addedByUid: map['addedByUid'] as String? ?? '',
      addedByName: map['addedByName'] as String? ?? '',
      approvalStatus: ApprovalStatus.fromString(
        map['approvalStatus'] as String?,
      ),
      approvedByUid: map['approvedByUid'] as String? ?? '',
      approvedByName: map['approvedByName'] as String? ?? '',
      approvedAt: map['approvedAt'] as String? ?? '',
    );
  }

  Map<String, dynamic> toMap() => {
        'salikId': salikId,
        'name': name,
        'fatherName': fatherName,
        'mobileNumber': mobileNumber,
        'whatsappNumber': whatsappNumber,
        'areaId': areaId,
        'address': address,
        'genderId': genderId,
        'bazamId': bazamId,
        'khanqahId': khanqahId,
        'salikCategoryId': salikCategoryId,
        'dateOfBaith': dateOfBaith,
        'referenceName': referenceName,
        'referenceMobile': referenceMobile,
        'isNafiAsbat': isNafiAsbat,
        'isSahibEMehfil': isSahibEMehfil,
        'nafiZikrId': nafiZikrId,
        'profilePicture': profilePicture,
        'createdDate': createdDate,
        'modifiedDate': modifiedDate,
        'isActive': isActive,
        'notes': notes,
        'addedByUid': addedByUid,
        'addedByName': addedByName,
        'approvalStatus': approvalStatus.toFirestore(),
        'approvedByUid': approvedByUid,
        'approvedByName': approvedByName,
        'approvedAt': approvedAt,
      };
}

File: app/src/main/java/com/example/salik_management_system/features/saliks/domain/model/Salik.kt
package com.example.salik_management_system.features.saliks.domain.model

import java.time.LocalDate

data class Salik(
    val salikId: String,
    val name: String,
    val fatherName: String,
    val mobileNumber: String,
    val whatsappNumber: String,
    val areaId: String,
    val address: String = "",
    val genderId: String,
    val bazamId: String = "",
    val dateOfBaith: String,
    val referenceName: String,
    val isNafiAsbat: Boolean = false,
    val isSahibEMehfil: Boolean = false,
    val createdDate: String,
    val modifiedDate: String,
    val isActive: Boolean = true,
    val addedByUid: String = "",
    val addedByName: String = "",
    val approvalStatus: ApprovalStatus = ApprovalStatus.Approved,
    val approvedByUid: String = "",
    val approvedByName: String = "",
    val approvedAt: String = "",
) {
    val isApproved: Boolean get() = approvalStatus == ApprovalStatus.Approved
    val isPending: Boolean get() = approvalStatus == ApprovalStatus.Pending
    val isRejected: Boolean get() = approvalStatus == ApprovalStatus.Rejected

    fun calculateAge(): Int? {
        val date = dateOfBaith.ifBlank { return null }
        return try {
            val baithDate = LocalDate.parse(dateOnly(date))
            val now = LocalDate.now()
            java.time.Period.between(baithDate, now).years
        } catch (e: Exception) {
            null
        }
    }

    private fun dateOnly(fullDate: String): String {
        return if (fullDate.contains("T")) {
            fullDate.substringBefore("T")
        } else if (fullDate.contains(" ")) {
            fullDate.substringBefore(" ")
        } else {
            fullDate
        }
    }

    fun toMap(): Map<String, Any?> = mapOf(
        "salikId" to salikId,
        "name" to name,
        "fatherName" to fatherName,
        "mobileNumber" to mobileNumber,
        "whatsappNumber" to whatsappNumber,
        "areaId" to areaId,
        "address" to address,
        "genderId" to genderId,
        "bazamId" to bazamId,
        "dateOfBaith" to dateOfBaith,
        "referenceName" to referenceName,
        "isNafiAsbat" to isNafiAsbat,
        "isSahibEMehfil" to isSahibEMehfil,
        "createdDate" to createdDate,
        "modifiedDate" to modifiedDate,
        "isActive" to isActive,
        "addedByUid" to addedByUid,
        "addedByName" to addedByName,
        "approvalStatus" to approvalStatus.toFirestore(),
        "approvedByUid" to approvedByUid,
        "approvedByName" to approvedByName,
        "approvedAt" to approvedAt,
    )

    companion object {
        fun fromMap(map: Map<String, Any?>, id: String? = null): Salik {
            fun text(key: String, legacyPrimary: String? = null, legacySecondary: String? = null): String {
                val primary = (map[key] as? String)?.trim().orEmpty()
                if (primary.isNotEmpty()) return primary
                val a = legacyPrimary?.let { (map[it] as? String)?.trim().orEmpty() }.orEmpty()
                val b = legacySecondary?.let { (map[it] as? String)?.trim().orEmpty() }.orEmpty()
                return when {
                    a.isNotEmpty() && b.isNotEmpty() && a != b -> "$a / $b"
                    a.isNotEmpty() -> a
                    else -> b
                }
            }

            fun bool(key: String): Boolean = when (val v = map[key]) {
                is Boolean -> v
                is String -> v.equals("true", ignoreCase = true)
                else -> false
            }

            fun dateOnly(fullDate: String): String {
                return if (fullDate.contains("T")) {
                    fullDate.substringBefore("T")
                } else if (fullDate.contains(" ")) {
                    fullDate.substringBefore(" ")
                } else {
                    fullDate
                }
            }

            return Salik(
                salikId = id ?: (map["salikId"] as? String).orEmpty(),
                name = text("name", "nameEnglish", "nameUrdu"),
                fatherName = text("fatherName", "fatherNameEnglish", "fatherNameUrdu"),
                mobileNumber = (map["mobileNumber"] as? String).orEmpty(),
                whatsappNumber = (map["whatsappNumber"] as? String).orEmpty(),
                areaId = ((map["areaId"] as? String) ?: (map["cityId"] as? String))?.trim().orEmpty(),
                address = (map["address"] as? String).orEmpty(),
                genderId = (map["genderId"] as? String) ?: "Male",
                bazamId = (map["bazamId"] as? String).orEmpty(),
                dateOfBaith = dateOnly((map["dateOfBaith"] as? String).orEmpty()),
                referenceName = (map["referenceName"] as? String).orEmpty(),
                isNafiAsbat = bool("isNafiAsbat"),
                isSahibEMehfil = bool("isSahibEMehfil"),
                createdDate = dateOnly((map["createdDate"] as? String).orEmpty()),
                modifiedDate = dateOnly((map["modifiedDate"] as? String).orEmpty()),
                isActive = map["isActive"]?.let {
                    when (it) {
                        is Boolean -> it
                        is String -> it.equals("true", ignoreCase = true)
                        else -> true
                    }
                } ?: true,
                addedByUid = (map["addedByUid"] as? String).orEmpty(),
                addedByName = (map["addedByName"] as? String).orEmpty(),
                approvalStatus = ApprovalStatus.fromString(map["approvalStatus"] as? String),
                approvedByUid = (map["approvedByUid"] as? String).orEmpty(),
                approvedByName = (map["approvedByName"] as? String).orEmpty(),
                approvedAt = dateOnly((map["approvedAt"] as? String).orEmpty()),
            )
        }
    }
}

enum class DuplicateSalikReason { Mobile, Name }

class DuplicateSalikException(val reason: DuplicateSalikReason) : Exception("Duplicate: $reason")

class SalikPermissionException(message: String) : Exception(message)

data class SalikDuplicateGroup(
    val id: String,
    val reasons: Set<DuplicateSalikReason>,
    val label: String,
    val saliks: List<Salik>,
)
